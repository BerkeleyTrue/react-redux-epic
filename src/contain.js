import { createElement, PureComponent } from 'react';

import debug from 'debug';
import invariant from 'invariant';

// // Using rtype signatures
// interface Action {
//   type: String,
//   payload?: Any,
//   ...meta?: Object
// }
//
// ActionCreator(...args?) => Action
//
// interface Options {
//   fetchAction?: ActionCreator,
//   getActionArgs?(props: Object, context: Object) => [],
//   isPrimed?(props: Object, context: Object) => Boolean,
//   shouldRefetch?(
//     props: Object,
//     nextProps: Object,
//     context: Object,
//     nextContext: Object
//   ) => Boolean,
// }
//
// interface contain {
//   (options?: Options, Component: ReactComponent) => ReactComponent
//   (options?: Object) => (Component: ReactComponent) => ReactComponent
// }


const log = debug('redux-epic:contain');

const isFunction = x => typeof x === 'function';

export default function contain(options = {}) {
  return Component => {
    const name = Component.displayName || 'Anon Component';
    let action;
    let isActionable = false;
    let hasRefetcher = isFunction(options.shouldRefetch);
    const getActionArgs = isFunction(options.getActionArgs) ?
      options.getActionArgs :
      (() => []);

    const isPrimed = isFunction(options.isPrimed) ?
      options.isPrimed :
      (() => false);

    function runAction(props, context, action) {
      const actionArgs = getActionArgs(props, context);
      invariant(
        Array.isArray(actionArgs),
        `
          ${name} getActionArgs should always return an array
          but got ${actionArgs}. check the render method of ${name}
        `
      );
      return action.apply(null, actionArgs);
    }


    class Container extends PureComponent {

      componentWillMount() {
        const { props, context } = this;
        const fetchAction = options.fetchAction;
        if (!options.fetchAction) {
          log(`Container(${name}) has no fetch action defined`);
          return;
        }
        if (isPrimed(this.props, this.context)) {
          log(`contain(${name}) is primed`);
          return;
        }

        action = props[options.fetchAction];
        isActionable = typeof action === 'function';

        invariant(
          isActionable,
          `
            ${fetchAction} should be a function on Container(${name})'s props
            but found ${action}. Check the fetch options for ${name}.
          `
        );

        runAction(
          props,
          context,
          action
        );
      }

      componentWillReceiveProps(nextProps, nextContext) {
        if (
          !isActionable ||
          !hasRefetcher ||
          !options.shouldRefetch(
            this.props,
            nextProps,
            this.context,
            nextContext
          )
        ) {
          return;
        }

        runAction(
          nextProps,
          nextContext,
          action
        );
      }

      render() {
        return createElement(
          Component,
          this.props
        );
      }
    }

    Container.displayName = `Container(${name})`;
    return Container;
  };
}
