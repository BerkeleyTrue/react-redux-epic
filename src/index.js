export { default as contain } from './contain';
export { default as renderToString } from './render-to-string';
export { default as render } from './render';
export { default as wrapRootEpic } from './wrap-root-epic';
