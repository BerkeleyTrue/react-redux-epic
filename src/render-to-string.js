import { Observable } from 'rxjs/Observable';
import { last } from 'rxjs/operator/last';
import { map } from 'rxjs/operator/map';
import ReactDOM from 'react-dom/server';
import debug from 'debug';

const log = debug('react-redux-epic:render-to-string');

// renderToString(
//   Component: ReactComponent,
//   epicMiddleware: EpicMiddleware
// ) => Observable[String]

export default function renderToString(element, wrappedEpic) {
  try {
    log('initial render pass started');
    ReactDOM.renderToStaticMarkup(element);
    log('initial render pass completed');
  } catch (e) {
    return Observable.throw(e);
  }
  log('calling action$ onCompleted');
  wrappedEpic.end();
  return Observable.create(wrappedEpic.subscribe)
    ::last(null, null, null)
    ::map(() => {
      wrappedEpic.restart();
      const markup = ReactDOM.renderToString(element);
      return { markup };
    });
}
