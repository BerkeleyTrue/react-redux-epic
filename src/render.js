import { render as _render, unmountComponentAtNode } from 'react-dom';
import { Observable } from 'rxjs/Observable';

// render(
//   element: ReactElement,
//   container: DOMElement,
// ) => Observable[RootInstance]

export default function render(element, container) {
  return Observable.create(observer => {
    try {
      _render(element, container, function() {
        observer.next(this);
      });
    } catch (e) {
      return observer.error(e);
    }

    return () => unmountComponentAtNode(container);
  });
}
