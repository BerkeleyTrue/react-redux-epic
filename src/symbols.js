export const $$getObservable = Symbol('@@getObservable');
export const $$end = Symbol('@@end');
export const $$restart = Symbol('@@restart');
export const $$isWrapped = Symbol('@@isWrapped');
