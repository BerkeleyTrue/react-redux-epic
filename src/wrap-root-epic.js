import 'rxjs';
import { EmptyObservable } from 'rxjs/observable/EmptyObservable';
import { Subject } from 'rxjs/Subject';
import { Subscriber } from 'rxjs/Subscriber';
import { EPIC_END } from 'redux-observable/lib/EPIC_END';

const endAction = { type: EPIC_END };

export default function wrapRootEpic(rootEpic) {
  let actionsProxy = EmptyObservable.create();
  let lifecycle = EmptyObservable.create();
  let subscription;
  let start;
  function wrappedEpic(actions, ...rest) {
    const results = new Subject();
    start = () => {
      subscription = new Subscriber();
      actionsProxy = new Subject();
      // how can subject inherit from ActionsObservable
      actionsProxy.ofType = actions.ofType;
      lifecycle = new Subject();
      const actionsSubscription = actions.subscribe(actionsProxy);
      const epicsSubscription = rootEpic(actionsProxy, ...rest)
        .subscribe(
          action => results.next(action),
          err => { throw err; },
          () => lifecycle.complete()
        );

      subscription.add(epicsSubscription);
      subscription.add(actionsSubscription);
    };
    start();
    return results;
  }

  wrappedEpic.subscribe =
    (...args) => lifecycle.subscribe.apply(lifecycle, args);
  wrappedEpic.unsubscribe = () => subscription.unsubscribe();
  wrappedEpic.end = () => {
    actionsProxy.next(endAction);
    actionsProxy.complete();
  };
  wrappedEpic.restart = () => {
    wrappedEpic.unsubscribe();
    actionsProxy.unsubscribe();
    start();
  };

  return wrappedEpic;
}
