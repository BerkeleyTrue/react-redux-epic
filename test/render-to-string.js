/* eslint-disable react/no-multi-comp */
/* eslint-disable react/prop-types */
import test from 'ava';
import { createStore, applyMiddleware } from 'redux';
import { Provider, connect } from 'react-redux';
import { createEpicMiddleware } from 'redux-observable';
import { Observable } from 'rxjs';
import { $$observable } from 'rxjs/symbol/observable';
import { Component, createElement } from 'react';

import { renderToString, wrapRootEpic } from '../src';

test('renderToString', t => {
  t.is(
    typeof renderToString,
    'function',
    'renderToString is a function'
  );
  t.pass('no tests yet');
});

test('renderToString(element, wrappedEpic) => Observable',
  t => {
    const element = createElement('div', null, 'hello world');
    const wrappedEpic = wrapRootEpic(actions => actions.ofType('FOO'));
    const renderer = renderToString(element, wrappedEpic);
    t.is(typeof renderer[$$observable], 'function');
  }
);

test('renderToString() => Observable[Error]',
  t => {
    return renderToString()
      .catch(
        err => {
          t.regex(
            err.message,
            /renderToString expects a valid react element/
          );
          return Observable.empty();
        }
      );
  }
);

test(
  'renderToString(ReactElement) => Observable[Error]',
  t => {
    const element = createElement('div', null, 'hello world');
    return renderToString(element)
      .catch(err => {
        t.regex(
          err.message,
          /renderToString expects a wrapped root epic/
        );
        return Observable.empty();
      });
  }
);

test(
  'renderToString(ReactElement, () => {}) => Observable[Error]',
  t => {
    const element = createElement('div', null, 'hello world');
    return renderToString(element, (actions) => actions.ofType('FOO'))
      .catch(err => {
        t.regex(
          err.message,
          /renderToString expects a wrapped root epic/
        );
        return Observable.empty();
      });
  }
);

test('renderToString(element, wrappedEpic) completes',
  t => {
    const element = createElement('div', null, 'hello world');
    const wrappedEpic = wrapRootEpic(actions => Observable
      .interval(1000)
      .mapTo({ type: 'FOO' })
      .takeUntil(actions.last())
    );
    createStore(
      x => x,
      applyMiddleware(createEpicMiddleware(wrappedEpic))
    );
    return renderToString(element, wrappedEpic)
      .do(({ markup }) => {
        t.is(typeof markup, 'string');
      })
      .last();
  }
);

test(
  'renderToString(thrower: ReactElement, wrappedEpic) => Observable[Error]',
  t => {
    class Thrower extends Component {
      componentWillMount() {
        throw new Error('Thrower throws');
      }
      render() {
        return createElement('div', null, 'Hello World');
      }
    }
    const element = createElement(Thrower);
    const wrappedEpic = wrapRootEpic(actions => Observable
      .interval(1000)
      .mapTo({ type: 'FOO' })
      .takeUntil(actions.last())
    );
    createStore(
      x => x,
      applyMiddleware(createEpicMiddleware(wrappedEpic))
    );
    return renderToString(element, wrappedEpic)
      .catch(err => {
        t.regex(
          err.message,
          /Thrower throws/
        );
        return Observable.empty();
      });
  }
);

test('renderToString with fetch completes without error',
  t => {
    class Content extends Component {
      componentWillMount() {
        this.props.dispatch({ type: 'FETCH' });
      }
      render() {
        const { content } = this.props;
        return createElement(
          'div',
          null,
          content
        );
      }
    }
    const rootEpic = actions => actions.ofType('FETCH')
      .switchMap(() => Observable.of({
          type: 'FETCH_COMPLETE',
          payload: 'WASSUP!!!'
        })
        .delay(1000)
    );
    function reducer(state = { content: 'fail' }, action) {
      if (action.type === 'FETCH_COMPLETE') {
        return {
          content: action.payload
        };
      }
      return state;
    }
    const wrappedEpic = wrapRootEpic(rootEpic);
    const store = createStore(
      reducer,
      applyMiddleware(createEpicMiddleware(wrappedEpic))
    );
    const element = createElement(
      Provider,
      { store },
      createElement(
        connect(state => state)(Content),
        null
      )
    );
    return renderToString(element, wrappedEpic)
      .do(({ markup }) => {
        t.is(typeof markup, 'string');
        t.notRegex(markup, /fail/);
        t.regex(markup, /wassup/i);
        t.snapshot({ markup });
        t.snapshot(store.getState());
      });
  }
);
