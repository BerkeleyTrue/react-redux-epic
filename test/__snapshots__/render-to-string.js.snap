// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderToString with fetch completes without error 1`] = `
Object {
  "markup": "<div data-reactroot=\\"\\" data-reactid=\\"1\\" data-react-checksum=\\"179442080\\">WASSUP!!!</div>",
}
`;

exports[`renderToString with fetch completes without error 2`] = `
Object {
  "content": "WASSUP!!!",
}
`;
